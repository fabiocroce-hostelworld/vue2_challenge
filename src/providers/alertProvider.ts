import { provide, inject, reactive, computed, ComputedRef } from "@vue/composition-api"

export enum alertTypes {
  SUCCESS = "success",
  ERROR = "error",
  INFO = "info",
  WARNING = "warning"
}

interface Alert {
  type: alertTypes
  message: string
  timerId?: number | undefined
}

export interface AlertProvider {
  addErrorAlert: (message: string) => void
  addSuccessAlert: (message: string) => void
  clearAlerts: () => void
  clearAlert: (idx: number) => void
  alerts: ComputedRef<Alert[]>
}

type alertConfig = { timeout?: number }

// Store for alert messages
export const setupAlertProvider = (config: alertConfig = {}): NonNullable<AlertProvider> => {
  const alertsArray = reactive<{ alerts: Array<Alert> }>({ alerts: [] })

  const clearByTimerId = (removeTimerId: number) => {
    alertsArray.alerts = alertsArray.alerts.filter(({ timerId }) => timerId !== removeTimerId)
  }

  const addAlert = ({ message, type }: Alert) => {
    if (config.timeout) {
      const timerId = setTimeout(() => {
        clearByTimerId(timerId)
      }, config.timeout)

      alertsArray.alerts.push({ message, type, timerId })
      return
    }
    alertsArray.alerts.push({ message, type })
  }

  const addErrorAlert = (message: string) => addAlert({ message, type: alertTypes.ERROR })
  const addSuccessAlert = (message: string) => addAlert({ message, type: alertTypes.SUCCESS })

  const clearAlerts = () => void (alertsArray.alerts.length = 0)
  const clearAlert = (removeIdx: number) => {
    alertsArray.alerts = alertsArray.alerts.filter((_, idx) => removeIdx !== idx)
  }
  const alerts = computed(() => alertsArray.alerts)

  return {
    addErrorAlert,
    addSuccessAlert,
    clearAlerts,
    clearAlert,
    alerts
  }
}

export const alertRef = Symbol("alert")

export const provideAlert = /* istanbul ignore next */ (config: alertConfig = {}) =>
  provide(alertRef, setupAlertProvider(config))
export const useAlert = () => inject<AlertProvider>(alertRef)
