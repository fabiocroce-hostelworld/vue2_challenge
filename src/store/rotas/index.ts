import { getRotaServices } from "@/api/clavaAPI"
import { IRotas } from "@/api/clavaAPI/rota/RotasResponseDTO"
import { IRotasList } from "@/api/clavaAPI/rotasUsers/RotasResponseDTO"
import { addNamespacePrefix } from "@/utils/vuexHelpers"

const types = {
  getters: {
    GET_ROTAS: "GET_ROTAS"
  },
  mutations: {
    STORE_ROTAS: "STORE_ROTAS",
    STORE_ROTA: "STORE_ROTA"
  },
  actions: {
    FETCH_ROTA: "fetchRota",
    GENERATE_ROTA: "generateRota"
  }
}

export const rotasTypes = addNamespacePrefix(types, "rotas")

type RotasState = {
  rotas: IRotasList
}

export default {
  namespaced: true,
  state: (): RotasState => ({
    rotas: [] as IRotasList
  }),
  getters: {
    [types.getters.GET_ROTAS](state: RotasState) {
      return state.rotas
    }
  },
  actions: {
    [types.actions.FETCH_ROTA]: async ({ commit }: { commit: Function }, rotaId: number) => {
      const rota = await getRotaServices().fetchRota(rotaId)

      commit(types.mutations.STORE_ROTA, rota)
    },
    [types.actions.GENERATE_ROTA]: async ({ dispatch }: { dispatch: Function }, date: string) => {
      const rotaId = await getRotaServices().generateRota(date)

      await dispatch(types.actions.FETCH_ROTA, rotaId)
    }
  },
  mutations: {
    [types.mutations.STORE_ROTAS]: (state: RotasState, rotas: IRotasList) => {
      state.rotas = rotas
    },
    [types.mutations.STORE_ROTA]: (state: RotasState, rota: IRotas) => {
      state.rotas.push(rota)
    }
  }
}
