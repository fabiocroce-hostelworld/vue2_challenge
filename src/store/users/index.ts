import { IUsersList } from "@/api/clavaAPI/rotasUsers/RotasResponseDTO"
import { addNamespacePrefix } from "@/utils/vuexHelpers"

const types = {
  getters: {
    GET_USERS: "getUsers",
    GET_MAPPED_USERS: "getMappedUsers",
    GET_USER_NAME: "getUserName"
  },
  mutations: {
    STORE_USERS: "STORE_USERS"
  },
  actions: {
    FETCH_USERS: "fetchUsers"
  }
}

export const usersTypes = addNamespacePrefix(types, "users")

type UsersState = {
  users: IUsersList
}

export default {
  namespaced: true,
  state: (): UsersState => ({
    users: [] as IUsersList
  }),
  getters: {
    [types.getters.GET_USERS](state: UsersState) {
      return state.users
    },
    [types.getters.GET_MAPPED_USERS](state: UsersState) {
      return Object.assign({}, ...state.users.map(user => ({ [user.userId]: user.user })))
    },
    [types.getters.GET_USER_NAME](_: UsersState, getters: { [x: string]: IUsersList }) {
      return (userId: number) => getters[types.getters.GET_MAPPED_USERS][userId]
    }
  },
  mutations: {
    [types.mutations.STORE_USERS]: (state: UsersState, users: IUsersList) => {
      state.users = users
    }
  }
}
