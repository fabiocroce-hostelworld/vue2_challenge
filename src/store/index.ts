import { getRotasUsersServices } from "@/api/clavaAPI"
import Vue from "vue"
import Vuex, { Store } from "vuex"
import rotas, { rotasTypes } from "@/store/rotas"
import users, { usersTypes } from "@/store/users"

Vue.use(Vuex)

export const rootTypes = {
  actions: {
    FETCH_ROTAS_AND_USERS: "fetchRotasAndUsers"
  }
}

export { rotasTypes, usersTypes }

export const storeConfig = {
  modules: {
    rotas,
    users
  },
  actions: {
    [rootTypes.actions.FETCH_ROTAS_AND_USERS]: async ({ commit }: { commit: Function }) => {
      const { users, rotas } = await getRotasUsersServices().fetchUsersAndRotas()

      commit(rotasTypes.mutations.STORE_ROTAS, rotas)
      commit(usersTypes.mutations.STORE_USERS, users)
    }
  }
}

export default new Store(storeConfig)
