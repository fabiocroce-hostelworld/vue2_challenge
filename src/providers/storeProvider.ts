import { provide, inject, computed, ComputedRef } from "@vue/composition-api"
import store from "@/store"

export interface StoreProvider {
  store: typeof store
  reactiveGetter: <T>(name: string) => ComputedRef<T>
}

// Vuex adapter for composition API
export const setupStoreProvider = (localStore: typeof store): StoreProvider => ({
  store: localStore,
  reactiveGetter: <T>(getterName: string): ComputedRef<T> => {
    return computed(() => localStore.getters[getterName])
  }
})

export const storeRef = Symbol("vuex-store")
export const provideStore = /* istanbul ignore next */ () => provide(storeRef, setupStoreProvider(store))
export const useStore = () => inject<StoreProvider>(storeRef)
