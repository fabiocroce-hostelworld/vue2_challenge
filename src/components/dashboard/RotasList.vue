<template>
  <v-card>
    <v-expansion-panels v-if="selectedEvents.length" multiple accordion class="m-4">
      <v-expansion-panel v-for="(events, idx) in selectedEvents" :key="idx" class="rota-dates px-5" expand>
        <v-expansion-panel-header class="rota-date"> {{ events[0] }}</v-expansion-panel-header>
        <v-expansion-panel-content v-if="events[1][ShiftTypes.MORNING] && selectedUsers.has(events[1][ShiftTypes.MORNING])">
          <v-chip class="secondary mr-2">Morning</v-chip
          ><span class="morning">{{ getUserNameById(events[1][ShiftTypes.MORNING]) }}</span>
        </v-expansion-panel-content>
        <v-expansion-panel-content
          v-if="events[1][ShiftTypes.AFTERNOON] && selectedUsers.has(events[1][ShiftTypes.AFTERNOON])"
        >
          <v-chip class="primary mr-2">Afternoon</v-chip
          ><span class="afternoon">{{ getUserNameById(events[1][ShiftTypes.AFTERNOON]) }}</span>
        </v-expansion-panel-content>
      </v-expansion-panel>
    </v-expansion-panels>
    <info v-else class="info" message="Select users to display the shifts" />
  </v-card>
</template>

<script lang="ts">
import { usersTypes } from "@/store"
import { computed, defineComponent, PropType } from "@vue/composition-api"
import { rotasTypes } from "@/store"
import { useStore } from "@/providers/storeProvider"
import { IRota, ShiftTypes } from "@/api/clavaAPI/rota/RotasResponseDTO"
import { IRotasList } from "@/api/clavaAPI/rotasUsers/RotasResponseDTO"
import Info from "@/components/helpers/Info.vue"

interface Props {
  selectedUsers: Set<number>
}

type MappedDay = {
  [type in ShiftTypes]?: number
}

type MappedRota = Record<string, MappedDay>

type MappedEvents = Array<[string, MappedDay]>

export default defineComponent<Props>({
  name: "RotasList",
  components: {
    Info
  },
  props: {
    selectedUsers: {
      type: Set as PropType<Set<number>>,
      required: true
    }
  },
  setup(props) {
    const { store, reactiveGetter } = useStore()!
    const rotas = reactiveGetter<IRotasList>(rotasTypes.getters.GET_ROTAS)

    const getUserNameById = (userId: number) => store.getters[usersTypes.getters.GET_USER_NAME](userId)

    const mappedEvents = computed<MappedEvents>(() => {
      return rotas.value.flatMap(thisRota =>
        Object.entries(
          thisRota.rota.reduce((acc: MappedRota, { date, userId, type }: IRota) => {
            acc[date] = acc[date] ?? {}

            acc[date][type] = userId

            return acc
          }, {})
        )
      )
    })

    const selectedEvents = computed(() =>
      mappedEvents.value.filter(mappedEvent => {
        const mornminguser = mappedEvent[1][ShiftTypes.MORNING]
        const afternoonUser = mappedEvent[1][ShiftTypes.AFTERNOON]

        return (
          (mornminguser && props.selectedUsers.has(mornminguser)) ||
          (afternoonUser && props.selectedUsers.has(afternoonUser))
        )
      })
    )

    return {
      selectedEvents,
      getUserNameById,
      ShiftTypes
    }
  }
})
</script>
