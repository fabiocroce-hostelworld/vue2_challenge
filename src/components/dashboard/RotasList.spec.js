import Vuex from "vuex"
import { shallowMount, createLocalVue } from "@vue/test-utils"
import { storeConfig, rotasTypes, usersTypes } from "@/store"
import RotasList from "@/components/dashboard/RotasList"
import rotaUsersMock from "@/api/clavaAPI/rotasUsers/rotasResponse.mock"
import { setupStoreProvider, storeRef } from "@/providers/storeProvider"

describe("RotasList", () => {
  let input
  let localVue
  let store

  beforeEach(() => {
    localVue = createLocalVue()
    store = new Vuex.Store(storeConfig)

    store.commit(rotasTypes.mutations.STORE_ROTAS, rotaUsersMock.rotas)
    store.commit(usersTypes.mutations.STORE_USERS, rotaUsersMock.users)

    input = {
      localVue,
      provide: {
        [storeRef]: setupStoreProvider(store)
      }
    }
  })

  it("Snapshot test", () => {
    input.propsData = {
      selectedUsers: new Set([2, 3])
    }
    const wrapper = shallowMount(RotasList, input)
    expect(wrapper.findAll(".info").length).toBe(0)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("Snapshot test - with warning empty users", () => {
    input.propsData = {
      selectedUsers: new Set()
    }
    const wrapper = shallowMount(RotasList, input)
    expect(wrapper.findAll(".info").length).toBe(1)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("expects events to be empty if selectedUsers has no user ids", () => {
    input.propsData = {
      selectedUsers: new Set()
    }
    const wrapper = shallowMount(RotasList, input)

    const panel = wrapper.findAll(".rota-dates")

    expect(panel.length).toEqual(0)
  })

  it("expects events not to be empty if selectedUsers has a userID (Titan Cox)", () => {
    input.propsData = {
      selectedUsers: new Set([5])
    }
    const wrapper = shallowMount(RotasList, input)

    const panel = wrapper.findAll(".rota-dates")

    expect(panel.length).toEqual(3)

    expect(panel.at(0).find(".rota-date").text()).toBe("2020-11-05")
    expect(panel.at(0).find(".morning").text()).toBe("Titan Cox")
    expect(panel.at(0).findAll(".afternoon").length).toBe(0)

    expect(panel.at(1).find(".rota-date").text()).toBe("2020-11-14")
    expect(panel.at(1).find(".afternoon").text()).toBe("Titan Cox")
    expect(panel.at(1).findAll(".morning").length).toBe(0)

    expect(panel.at(2).find(".rota-date").text()).toBe("2020-11-16")
    expect(panel.at(2).find(".morning").text()).toBe("Titan Cox")
    expect(panel.at(2).findAll(".afternoon").length).toBe(0)
  })
})
