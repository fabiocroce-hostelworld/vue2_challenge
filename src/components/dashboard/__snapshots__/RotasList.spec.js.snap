// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RotasList Snapshot test - with warning empty users 1`] = `
<v-card-stub loaderheight="4" tag="div">
  <info-stub message="Select users to display the shifts" class="info"></info-stub>
</v-card-stub>
`;

exports[`RotasList Snapshot test 1`] = `
<v-card-stub loaderheight="4" tag="div">
  <v-expansion-panels-stub activeclass="v-item--active" multiple="true" accordion="true" class="m-4">
    <v-expansion-panel-stub expand="" class="rota-dates px-5">
      <v-expansion-panel-header-stub expandicon="$expand" class="rota-date"> 2020-11-02</v-expansion-panel-header-stub>
      <v-expansion-panel-content-stub>
        <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="secondary mr-2">Morning</v-chip-stub><span class="morning">Ann Doe</span>
      </v-expansion-panel-content-stub>
      <!---->
    </v-expansion-panel-stub>
    <v-expansion-panel-stub expand="" class="rota-dates px-5">
      <v-expansion-panel-header-stub expandicon="$expand" class="rota-date"> 2020-11-03</v-expansion-panel-header-stub>
      <v-expansion-panel-content-stub>
        <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="secondary mr-2">Morning</v-chip-stub><span class="morning">Nicole Woo</span>
      </v-expansion-panel-content-stub>
      <!---->
    </v-expansion-panel-stub>
    <v-expansion-panel-stub expand="" class="rota-dates px-5">
      <v-expansion-panel-header-stub expandicon="$expand" class="rota-date"> 2020-11-06</v-expansion-panel-header-stub>
      <!---->
      <v-expansion-panel-content-stub>
        <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="primary mr-2">Afternoon</v-chip-stub><span class="afternoon">Nicole Woo</span>
      </v-expansion-panel-content-stub>
    </v-expansion-panel-stub>
    <v-expansion-panel-stub expand="" class="rota-dates px-5">
      <v-expansion-panel-header-stub expandicon="$expand" class="rota-date"> 2020-11-07</v-expansion-panel-header-stub>
      <!---->
      <v-expansion-panel-content-stub>
        <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="primary mr-2">Afternoon</v-chip-stub><span class="afternoon">Ann Doe</span>
      </v-expansion-panel-content-stub>
    </v-expansion-panel-stub>
    <v-expansion-panel-stub expand="" class="rota-dates px-5">
      <v-expansion-panel-header-stub expandicon="$expand" class="rota-date"> 2020-11-09</v-expansion-panel-header-stub>
      <!---->
      <v-expansion-panel-content-stub>
        <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="primary mr-2">Afternoon</v-chip-stub><span class="afternoon">Nicole Woo</span>
      </v-expansion-panel-content-stub>
    </v-expansion-panel-stub>
    <v-expansion-panel-stub expand="" class="rota-dates px-5">
      <v-expansion-panel-header-stub expandicon="$expand" class="rota-date"> 2020-11-12</v-expansion-panel-header-stub>
      <!---->
      <v-expansion-panel-content-stub>
        <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="primary mr-2">Afternoon</v-chip-stub><span class="afternoon">Ann Doe</span>
      </v-expansion-panel-content-stub>
    </v-expansion-panel-stub>
    <v-expansion-panel-stub expand="" class="rota-dates px-5">
      <v-expansion-panel-header-stub expandicon="$expand" class="rota-date"> 2020-11-13</v-expansion-panel-header-stub>
      <!---->
      <v-expansion-panel-content-stub>
        <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="primary mr-2">Afternoon</v-chip-stub><span class="afternoon">Nicole Woo</span>
      </v-expansion-panel-content-stub>
    </v-expansion-panel-stub>
    <v-expansion-panel-stub expand="" class="rota-dates px-5">
      <v-expansion-panel-header-stub expandicon="$expand" class="rota-date"> 2020-11-18</v-expansion-panel-header-stub>
      <v-expansion-panel-content-stub>
        <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="secondary mr-2">Morning</v-chip-stub><span class="morning">Nicole Woo</span>
      </v-expansion-panel-content-stub>
      <!---->
    </v-expansion-panel-stub>
    <v-expansion-panel-stub expand="" class="rota-dates px-5">
      <v-expansion-panel-header-stub expandicon="$expand" class="rota-date"> 2020-11-19</v-expansion-panel-header-stub>
      <v-expansion-panel-content-stub>
        <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" class="secondary mr-2">Morning</v-chip-stub><span class="morning">Ann Doe</span>
      </v-expansion-panel-content-stub>
      <!---->
    </v-expansion-panel-stub>
  </v-expansion-panels-stub>
</v-card-stub>
`;
